### PR TITLE
Refactored cfbs add and cfbs.json handling

### PR DESCRIPTION
**Pull request by olehermanse**

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thank you for submitting a pull request! Maybe @vpodzime can review this?

File: cfbs/index.py [hunk 1]
```diff
@@ -291,3 +292,30 @@ def add(self, module, remote_config=None, dependent=None):
         else:
             print("Added module: %s" % module["name"])
         self.validate_added_module(module)
+
+
+class CFBSConfig:
```
Comment (larsewi) [inline cfbs/index.py]: Maybe this should be a lazily initialized singleton class inheriting `CFBSJson`?

File: cfbs/core.py [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+        return self[key]
+
+    def _expand_index(self):
+        index = self._unexpanded
+        if type(index) in (dict, OrderedDict):
+            self._data = {"type": "index", "index": index}
+            return
+
+        assert type(index) is not Index
+        assert type(index) is str
```
Comment (larsewi) [inline cfbs/core.py:114]: `type(index) is str` implies that `type(index) is not Index`, thus  `type(index) is not Index` is excessive

File: test/shell/011_update.sh [hunk 3]
```diff
@@ -5,31 +5,62 @@ mkdir -p ./tmp/
...
 
-cfbs update
-# This test currently just checks that cfbs update runs succesfully,
-# and doesn't break the cfbs.json project config file
-# TODO: Expand this test once we can add a specific version, and
-#       check that update actually changes something
+# This is not perfect, it relies on the JSON formatting, and it doesn't check
+# everything it could. Still, it tests a lot and was really easy to add.
+# Also, it should not break on new commits(!)
+# TODO: Use jq, a python script, cfbs validate or soething similar
```
Comment (craigcomstock) [inline test/shell/011_update.sh:30]: `s/soething/something/` :)

**Review by craigcomstock — COMMENTED** [review]
Looks great! One typo in a comment is all I found to fix. :)

File: cfbs/core.py [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+def _has_autorun_tag(filename):
+    assert os.path.isfile(filename)
+    content = read_file(filename)
+
+    return (
+        "meta:" in content
+        and "tags" in content
+        and "slist" in content
+        and "autorun" in content
+    )
```
Comment (larsewi) [inline cfbs/core.py:197]: What happens if i have the following in a policy file?
```
# TODO: add autorun

bundle agent bogus
{
  meta:
      "tags" slist => { "deprecated" };
}
```

Comment (larsewi) [reply cfbs/core.py:197]: This is probably outside the scope of this PR, but we should make a ticket for this if it is a problem.

Comment (olehermanse) [reply cfbs/core.py:197]: Yes, we don't want to implement policy parsing, so there are many ways to circumvent this check.

File: cfbs/core.py [hunk 5]
```diff
@@ -91,10 +91,18 @@ def __init__(self, index=_DEFAULT_INDEX):
...
-        return self.get_modules()[key]
+        return self.data["index"][key]
+
+    def items(self):
+        return self.data["index"].items()
+
+    def get(self, key, default=None):
+        if not key in self:
+            return default
+        return self[key]
```
Comment (vpodzime) [inline cfbs/core.py:105]: I'd do `return self.data.get(key, default)`, but YMMV.

File: cfbs/core.py [hunk 6]
```diff
@@ -375,60 +371,95 @@ def _add_modules(
...
+        exclude_names = [m["name"] for m in exclude]
+        index = self.index
+        dependencies = []
+        for module in modules:
+            if "dependencies" in module:
+                for dep in module["dependencies"]:
+                    if dep not in exclude_names:
+                        m = index.get_module_object(dep, module["name"])
+                        dependencies.append(m)
+                        exclude_names.append(dep)
```
Comment (vpodzime) [inline cfbs/core.py:412]: Might be nicer and more "pythonic" to check the opposite conditions and do `continue` to reduce nesting here.

File: cfbs/commands.py [hunk 7]
```diff
@@ -372,11 +372,15 @@ def update_command(non_interactive=False):
...
-                        + "Do you want to use the new build steps? [y/N]"
+                    ans = (
+                        input(
+                            "Module %s has different build steps now\n" % module["name"]
+                            + "old steps: %s\n" % module["steps"]
+                            + "new steps: %s\n" % index_info["steps"]
+                            + "Do you want to use the new build steps? [y/N]"
+                        )
+                        if not non_interactive
+                        else "yes"
```
Comment (vpodzime) [inline cfbs/commands.py:383]: Good old 
```
if non_interactive:
  ans = "yes"
else:
  ans = input(...)
```
would be better here, AFAICT.

**Review by vpodzime — APPROVED** [review]
Looks good to me. Nice improvements! I only have a couple YMMV things. And I wonder if this fixes #91 ?

**Comment by olehermanse** [comment]
Since there are no blockers I will merge and address comments in a follow-up PR.

File: cfbs/core.py (hunk 2 above)
Comment (olehermanse) [reply cfbs/core.py:114]: Thanks, fixed: https://github.com/cfengine/cfbs/pull/93/commits/f350c7e37d1c3b19d7b2a7b2bb7ebbc2a551f9eb

File: test/shell/011_update.sh (hunk 3 above)
Comment (olehermanse) [reply test/shell/011_update.sh:30]: Thanks, fixed: https://github.com/cfengine/cfbs/pull/93/commits/3126116cb6d1fb079a90b43069c529b380dbbfaa

File: cfbs/core.py (hunk 5 above)
Comment (olehermanse) [reply cfbs/core.py:105]: Fixed: https://github.com/cfengine/cfbs/pull/93/commits/777c99684c7ab45ed6cd1806699c8b180e659c8e

File: cfbs/core.py (hunk 6 above)
Comment (olehermanse) [reply cfbs/core.py:412]: Yep, thanks, reduced nesting here: https://github.com/cfengine/cfbs/pull/93/commits/c0d37d84e08518893d4950b3264f58a83e542ab0

File: cfbs/commands.py (hunk 7 above)
Comment (olehermanse) [reply cfbs/commands.py:383]: Yep, agreed: https://github.com/cfengine/cfbs/pull/93/commits/75dba0e8df1c18db8933c4b13f8653ebdd02f588

**Comment by olehermanse** [comment]
> And I wonder if this fixes #91 ?

@vpodzime it does not.

**Pull request merged**